### PR TITLE
IDEMPIERE-4606: Adding support for generic method to get model using …

### PR DESCRIPTION
**Pull request by dpansheriya**
…UUID

### CODE REVIEW

File: org.adempiere.base/src/org/adempiere/base/DefaultModelFactory.java [hunk 1]
```diff
@@ -322,4 +322,14 @@ public PO getPO(String tableName, ResultSet rs, String trxName) {
 			s_log.log(Level.SEVERE, "(rs) - Not found - Table=" + tableName);
 		return null;
 	}
+
+	@Override
+	public PO getPO(String tableName, String uuID, String trxName)
```
Comment (hengsin) [inline org.adempiere.base/src/org/adempiere/base/DefaultModelFactory.java:327]: this should be moved to the interface as a default method. very unlikely that any model factory implementation will need to override this.

File: org.adempiere.base/src/org/compiere/model/MTable.java [hunk 2]
```diff
@@ -601,6 +601,58 @@ public PO getPO (ResultSet rs, String trxName)
...
+
+		if (po == null)
+		{
+			// TODO have to check working.
+			po = new GenericPO(tableName, getCtx(), 0, trxName);
+		}
+
+		return po;
+	} // getPOByUUID
+
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/MTable.java:655]: this implementation plus the addition of new method to IModelFactory interface is actually pointless - it is just a unnecessary long way to reach the same necessary path of MTable.getPO(0, trxName) > po.loadByUU(uuid, trxName)

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 3]
```diff
@@ -290,6 +319,10 @@ protected void copyPO(PO copy)
 	
 	/** Immutable flag **/
 	private boolean m_isImmutable = false;
+	
+	private String m_uuid = null;
+	/** UU flag **/
+	private boolean m_isUUID = false;
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:325]: should avoid adding another new flag here - increase potential for side effect and complicate po state management.

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 4]
```diff
@@ -1343,6 +1376,35 @@ protected void load (int ID, String trxName)
...
+		{
+			setKeyInfo();
+			load(trxName);
+		}
+		else
+		{
+			loadDefaults();
+			m_createNew = true;
+			setKeyInfo();
+			loadComplete(true);
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:1405]: should refactor and share the code instead of just duplicate this else part from the "load (int ID, String trxName)" method.

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 5]
```diff
@@ -1378,17 +1440,24 @@ public boolean load (String trxName)
 		try
 		{
 			pstmt = DB.prepareStatement(sql.toString(), m_trxName);	//	local trx only
-			for (int i = 0; i < m_IDs.length; i++)
+			if(m_isUUID)
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:1443]: don't need the flag here, just pass in the uuid parameter and check that for null is simpler and safer

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 6]
```diff
@@ -3173,6 +3253,18 @@ protected void saveNew_afterSetID()
 	public String get_WhereClause (boolean withValues)
 	{
 		StringBuilder sb = new StringBuilder();
+		
+		if (m_isUUID) {
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:3257]: same here, instead of flag, pass in the uuid param is simpler and safer.

File: org.adempiere.base/src/org/idempiere/model/IMappedModelFactory.java [hunk 7]
```diff
@@ -46,7 +46,8 @@
 	 */
 	void addMapping(String tableName, Supplier<Class<?>> classSupplier,
 			BiFunction<Integer, String, ? extends PO> recordIdFunction,
-			BiFunction<ResultSet, String, ? extends PO> resultSetFunction);
+			BiFunction<ResultSet, String, ? extends PO> resultSetFunction,
+			BiFunction<String, String, ? extends PO> recordUUFunction);
```
Comment (hengsin) [inline org.adempiere.base/src/org/idempiere/model/IMappedModelFactory.java:50]: the changes to IMappedModelFactory and the classes below is not needed as you have remove the uuid method from IModelFactory.

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 8]
```diff
@@ -291,6 +320,9 @@ protected void copyPO(PO copy)
 	/** Immutable flag **/
 	private boolean m_isImmutable = false;
 
+	/** Record UUID **/
+	private String m_uuid = null;
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:324]: I think it should be better if we remove this instance variable as well. the uuid can be parameter that's being pass around. If you are keeping this, you will have to ensure it always has the right uuid value, otherwise the PO state can be inconsistent at time.

**Comment by dpansheriya** [comment]
@hengsin I fixed as per your suggestion. Please review and appreciate your feedback.

File: org.adempiere.base/src/org/idempiere/model/IMappedModelFactory.java (hunk 7 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/idempiere/model/IMappedModelFactory.java]: @hengsin good catch, I missed it. Let me correct it.

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 8 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java:324]: @hengsin I tried to do this but looks get issue in get_WhereClause method as it invoked from many places and we have to pass null arguments on most of place. I think as currently the way it used without having getters and setters for m_uuid, we don't needs to maintain state of it. If we try to load value in it after reading record from DB or from RS, it may create more confusion as then after record always loaded by UU even ID present.

Comment (hengsin) [reply org.adempiere.base/src/org/compiere/model/PO.java:324]: I don't see why get_WhereClause is a problem, with get_WhereClause (boolean withValues, string uuid) and get_WhereClause (boolean withValues) { return get_WhereClause (boolean withValues, null);} should be sufficient maintain compatibility with existing code. Having instance variable in inconsistent state is always a potential source of unwanted side effect in future.

File: org.adempiere.base/src/org/idempiere/model/MappedModelFactory.java [hunk 9]
```diff
@@ -44,6 +44,7 @@
 public class MappedModelFactory implements IModelFactory, IMappedModelFactory {
 
 	private final ConcurrentHashMap<String, Supplier<Class<?>>> classMap = new ConcurrentHashMap<>();
+	private final ConcurrentHashMap<String, BiFunction<String, String, ? extends PO>> recordUUMap = new ConcurrentHashMap<>();
```
Comment (hengsin) [inline org.adempiere.base/src/org/idempiere/model/MappedModelFactory.java:47]: this should be removed

File: org.adempiere.base/src/org/idempiere/model/MappedModelFactory.java [hunk 10]
```diff
@@ -73,16 +74,19 @@ public PO getPO(String tableName, ResultSet rs, String trxName) {
...
 	}
 	
 	@Override
 	public void removeMapping(String tableName) {
 		classMap.remove(tableName);
 		recordIdMap.remove(tableName);
 		resultSetMap.remove(tableName);
+		recordUUMap.remove(tableName);
 	}
+
```
Comment (hengsin) [inline org.adempiere.base/src/org/idempiere/model/MappedModelFactory.java:91]: please revert the changes to this class

File: org.idempiere.test/src/org/idempiere/test/model/MappedModelFactoryTest.java [hunk 11]
```diff
@@ -65,8 +65,10 @@ public MappedModelFactoryTest() {
 	@Order(1)
 	public void testDefaultMappedModelFactory() {
 		IMappedModelFactory mappedFactory = Core.getMappedModelFactory();
-		mappedFactory.addMapping(MyTest.Table_Name, () -> MyTest.class, (id, trxName) -> new MyTest(Env.getCtx(), id, trxName), 
-				(rs, trxName) -> new MyTest(Env.getCtx(), rs, trxName));		
+		mappedFactory.addMapping(MyTest.Table_Name, () -> MyTest.class, 
+		                         (id, trxName) -> new MyTest(Env.getCtx(), id, trxName), 
+		                         (rs, trxName) -> new MyTest(Env.getCtx(), rs, trxName), 
+		                         (uuID, trxName) -> new MyTest(Env.getCtx(), uuID, trxName));
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/model/MappedModelFactoryTest.java:71]: please revert the changes to this class

**Comment by hengsin** [comment]
would be good if unit testing can be added to org.idempiere.test to verify the new method added to genericpo, po and mtable is working as expected.

File: org.adempiere.base/src/org/adempiere/base/IModelFactory.java [hunk 12]
```diff
@@ -46,4 +46,5 @@
 	 *	@return PO for Record or null
 	 */
 	public PO getPO (String tableName, ResultSet rs, String trxName);
+
```
Comment (hengsin) [inline org.adempiere.base/src/org/adempiere/base/IModelFactory.java:49]: would be good to remove this unnecessary white space only changes (same for the file above)

File: org.adempiere.base/src/org/compiere/model/MTable.java [hunk 13]
```diff
@@ -601,6 +601,22 @@ public PO getPO (ResultSet rs, String trxName)
...
+	 * Get PO Class Instance
+	 * 
+	 * @param  uuID    UUID
+	 * @param  trxName transaction
+	 * @return         PO for Record or null
+	 */
+	public PO getPOByUU (String uuID, String trxName)
+	{
+		PO po = getPO(0, trxName);
+		if (po != null)
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/MTable.java:614]: don't think you can ever get po==null here.

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 14]
```diff
@@ -1343,6 +1375,26 @@ protected void load (int ID, String trxName)
...
+		m_uuid = uuID;
+		checkImmutable();
+
+		if (log.isLoggable(Level.FINEST))
+			log.finest("uuID=" + uuID);
+		if (!Util.isEmpty(m_uuid, true))
+		{
+			load(trxName);
+		}
+	} // loadByUU
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:1397]: if the definition here is if uuid==null, it should create a new PO record, it should be simplify to just if (Util.isEmpty(uuid, true)) load(0, trxName); else load(uuid, trxName);

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 15]
```diff
@@ -1569,6 +1628,17 @@ else if (clazz == String.class)
...
 
+	/**
+	 * Load record with UUID
+	 * @param uuID - UUID
+	 * @param trxName - Transaction name
+	 */
+	protected void load(String uuID, String trxName)
+	{
+		load(0, trxName);
+		loadByUU(uuID, trxName);
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:1639]: would prefer the following instead, change the original load(trxName) method to load(String uuid, String trxName) and add load(String trxName) {load(null, trxName);}. Do the empty uuid check above in the rename load(uuid, trxname) method.

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 16]
```diff
@@ -3173,6 +3243,18 @@ protected void saveNew_afterSetID()
...
+		{
+			sb.append(getUUIDColumnName()).append("=");
+			if (withValues)
+				sb.append(m_uuid);
+			else
+				sb.append("?");
+
+			return sb.toString();
+		}
+
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:3257]: suggestion: rename this to get_whereclause(string uuid, boolean withValues). add get_whereclause(boolean withvalues) {return get_whereclause(null, withvalues);}

File: org.idempiere.test/src/org/idempiere/test/model/MappedModelFactoryTest.java [hunk 17]
```diff
@@ -65,8 +65,9 @@ public MappedModelFactoryTest() {
 	@Order(1)
 	public void testDefaultMappedModelFactory() {
 		IMappedModelFactory mappedFactory = Core.getMappedModelFactory();
-		mappedFactory.addMapping(MyTest.Table_Name, () -> MyTest.class, (id, trxName) -> new MyTest(Env.getCtx(), id, trxName), 
-				(rs, trxName) -> new MyTest(Env.getCtx(), rs, trxName));		
+		mappedFactory.addMapping(MyTest.Table_Name, () -> MyTest.class, 
+		                         (id, trxName) -> new MyTest(Env.getCtx(), id, trxName), 
+		                         (rs, trxName) -> new MyTest(Env.getCtx(), rs, trxName));
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/model/MappedModelFactoryTest.java:70]: please revert changes to this class too

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 8 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java:324]: @hengsin ohk Let me do this.

File: org.idempiere.test/src/org/idempiere/test/model/MappedModelFactoryTest.java (hunk 11 above)
Comment (dpansheriya) [reply org.idempiere.test/src/org/idempiere/test/model/MappedModelFactoryTest.java]: @hengsin I tried resetting but not getting any changes uncommitted.

File: org.adempiere.base/src/org/compiere/model/MTable.java (hunk 13 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/MTable.java:614]: Agree

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 15 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java:1639]: After refactoring, this method no longer required.

File: org.idempiere.test/src/org/idempiere/test/model/MappedModelFactoryTest.java (hunk 17 above)
Comment (dpansheriya) [reply org.idempiere.test/src/org/idempiere/test/model/MappedModelFactoryTest.java]: This is done in latest commit

Comment (dpansheriya) [reply org.idempiere.test/src/org/idempiere/test/model/MappedModelFactoryTest.java]: This is done in latest commit

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 16 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java:3264]: In refactored code, this is handled.

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 14 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java:1395]: This is done in latest commit

**Comment by hengsin** [comment]
@CarlosRuiz-globalqss, do  you think is necessary to re-generate all model classes for this ticket ?

**Comment by CarlosRuiz-globalqss** [comment]
> @CarlosRuiz-globalqss, do you think is necessary to re-generate all model classes for this ticket ?

Hi @hengsin - my comments in ticket

**Comment by hengsin** [comment]
> > @CarlosRuiz-globalqss, do you think is necessary to re-generate all model classes for this ticket ?
> 
> Hi @hengsin - my comments in ticket

Sorry, can't find your comments there that's related to the question I ask above.

**Comment by CarlosRuiz-globalqss** [comment]
redacting ... sorry for the delay

**Comment by hengsin** [comment]
hi @dpansheriya , as per Carlos's comments in the ticket, please remove allthe uuid constructor related changes from PO and X_Test.

**Comment by dpansheriya** [comment]
> hi @dpansheriya , as per Carlos's comments in the ticket, please remove all the uuid constructor related changes from PO and X_Test.

Sure, I am doing it. 
I am keeping getPOByUU as I am thinking to use return getPO(PO.getUUIDColumnName(getTableName())+"="+uuID, trxName); is not good as query is not parameterize.

File: org.adempiere.base/src/org/adempiere/model/GenericPO.java [hunk 18]
```diff
@@ -81,6 +81,16 @@ public GenericPO(String tableName, Properties ctx, ResultSet rs, String trxName)
...
+	/**
+	 * @param tableName
+	 * @param ctx
+	 * @param uuID 
+	 * @param trxName
+	 */
+	public GenericPO(String tableName, Properties ctx, String uuID, String trxName) {
+		super(new PropertiesWrapper(ctx, tableName), 0, trxName, null, uuID);
+	}
+
```
Comment (hengsin) [inline org.adempiere.base/src/org/adempiere/model/GenericPO.java:93]: this constructor should be removed as well

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 19]
```diff
@@ -190,6 +190,20 @@ public PO (Properties ctx, ResultSet rs, String trxName)
...
+	/**
+	 * Create & Load existing Persistent Object.
+	 * 
+	 * @param ctx     context
+	 * @param ID      the ID if 0, the record defaults are applied - ignored if re exists
+	 * @param trxName transaction name
+	 * @param rs      optional - load from current result set position (no navigation, not closed)
+	 * @param uuID    optional - load from uuID
+	 */
+	public PO(Properties ctx, int ID, String trxName, ResultSet rs, String uuID)
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:206]: this constructor should be removed as well

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 20]
```diff
@@ -1343,13 +1359,44 @@ protected void load (int ID, String trxName)
...
+
+		if (log.isLoggable(Level.FINEST))
+			log.finest("uuID=" + uuID);
+		if (!Util.isEmpty(uuID, true))
+		{
+			load(uuID,trxName);
+		}else {
+			load(0,trxName);
+		}
+	} // loadByUU
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:1382]: without the uuid constructor requirement, we just need to keep one of loadByUU(uuid, trxName) or load(uuid, trxName) (if we keep both, load(uuid, trxName) should be change to private or protected). and to keep things simple, for loadByUU(uuid, trxName), just return null if uuid is null or empty.

File: org.adempiere.base/src/org/compiere/model/MTable.java [hunk 21]
```diff
@@ -601,6 +601,21 @@ public PO getPO (ResultSet rs, String trxName)
 		return po;
 	}	//	getPO
 
+	/**
+	 * Get PO Class Instance
+	 * 
+	 * @param  uuID    UUID
+	 * @param  trxName transaction
+	 * @return         PO for Record or null
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/MTable.java:609]: this javadoc is wrong, it will never return null

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 20 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java:1382]: Shouldn't we allow null UU to create new record?

Comment (hengsin) [reply org.adempiere.base/src/org/compiere/model/PO.java:1382]: No, we don't need that  since the intention here is not to use uuid as primary key. The existing approach of id=0 should continue to be used to create new record.

File: org.adempiere.base/src/org/compiere/model/MTable.java (hunk 21 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/MTable.java]: @hengsin  you may referencing old commit as I did remove "or null" in commit 408fe45 (Latest one)

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 20 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java:1366]: This also done in latest commit.

File: org.adempiere.base/src/org/adempiere/model/GenericPO.java (hunk 18 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/adempiere/model/GenericPO.java]: Done

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 19 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java]: Done

File: org.adempiere.base/src/org/compiere/model/MTable.java [hunk 22]
```diff
@@ -601,6 +601,21 @@ public PO getPO (ResultSet rs, String trxName)
...
+	 * @return         PO for Record
+	 */
+	public PO getPOByUU (String uuID, String trxName)
+	{
+		PO po = getPO(0, trxName);
+		po.loadByUU(uuID, trxName);
+
+		return po;
+	} // getPOByUU
+
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/MTable.java:618]: Since we added new method to MTable, we need to update the serialVersionUID of MTable.

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 23]
```diff
@@ -1343,13 +1343,44 @@ protected void load (int ID, String trxName)
...
+		m_newValues = new Object[get_ColumnCount()];
+		checkImmutable();
+
+		if (log.isLoggable(Level.FINEST))
+			log.finest("uuID=" + uuID);
+		if (!Util.isEmpty(uuID, true))
+		{
+			load(uuID,trxName);
+		}else {
+			throw new IllegalArgumentException("Invalid null UU - Must pass valid UU");
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:1364]: the null check should be at the top. since we are throwing illegalargumentexception here, the method shouldn't do anything if uuid is null/empty.

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 24]
```diff
@@ -1343,13 +1343,44 @@ protected void load (int ID, String trxName)
...
 	 *  @param trxName transaction
 	 *  @return true if loaded
 	 */
-	public boolean load (String trxName)
+	public boolean load (String trxName) {
+		return load(null, trxName);
+	}
+	
+	/**
+	 *  (re)Load record with m_ID[*]
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:1378]: the javadoc should be updated to reflect the addition of uuID option.

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 25]
```diff
@@ -3164,15 +3202,37 @@ protected void saveNew_afterSetID()
...
+	 * 	@return where clause
+	 */
+	public String get_WhereClause (boolean withValues) {
+		return get_WhereClause(withValues,null);
+	}
 
 	/**
 	 * 	Create Single/Multi Key Where Clause
 	 * 	@param withValues if true uses actual values otherwise ?
+	 *  @param uuID RecordUU
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:3218]: javadoc should be updated to reflect the addtion of uuid option.

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 26]
```diff
@@ -3164,15 +3202,37 @@ protected void saveNew_afterSetID()
...
-	public String get_WhereClause (boolean withValues)
+	public String get_WhereClause (boolean withValues, String uuID)
 	{
 		StringBuilder sb = new StringBuilder();
+
+		if (!Util.isEmpty(uuID, true))
+		{
+			sb.append(getUUIDColumnName()).append("=");
+			if (withValues)
+				sb.append(uuID);
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:3229]: this is not correct - you need to  quote the uuID string value, i.e DB.TO_STRING(uuID)

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 25 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java:3218]: It is already there. Please note added line is for uuID paramater.

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 23 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java]: Done

File: org.adempiere.base/src/org/compiere/model/MTable.java (hunk 22 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/MTable.java:618]: Done

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 24 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java]: Fixed

File: org.adempiere.base/src/org/compiere/model/PO.java [hunk 27]
```diff
@@ -1343,13 +1343,44 @@ protected void load (int ID, String trxName)
...
+	 */
+	public void loadByUU(String uuID, String trxName)
+	{
+		// reset new values
+		m_newValues = new Object[get_ColumnCount()];
+		checkImmutable();
+
+		if (log.isLoggable(Level.FINEST))
+			log.finest("uuID=" + uuID);
+		if (Util.isEmpty(uuID, true))
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/PO.java:1360]: sorry, I means this should be at top, i.e comes before the "//reset new values" line.

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 26 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java]: Done

File: org.adempiere.base/src/org/compiere/model/PO.java (hunk 27 above)
Comment (dpansheriya) [reply org.adempiere.base/src/org/compiere/model/PO.java]: Done

**Review by hengsin — APPROVED** [review]
looks good.
